### PR TITLE
Qt: bug fixes and enhancement to traffic graph widget 

### PR DESCRIPTION
**Pull request by krychlicki**
All changes are connected to traffic graph widget.
Bug fixes:
- clear traffic graph widget after clicking the clear button
- set default sample height so after clearing graph has axis lines - currently it's just a black rectangle

Enhancement:
- having 288 graph ranges available is probably unnecessary (for example can't imagine that some uses range 13h 25m), so I limited it to only 10:
5m, 10m, 15m, 30m, 1h, 2h, 3h, 6h, 12h, 24h
If more ranges are needed please let me know 
- added possibility to change graph range without losing collected traffic samples
- data storage was designed to contain only necessary data to create graphs for all supported ranges
eg. 
for 10m range only half of 10m samples are stored - the second half is calculated from 5m range samples,
for 1h range data is created based on:
all 5m samples, 
2/3 of 15m samples,
1/2 of 30m samples,
1/2 of 1h samples
- all traffic graph related data and logic were encapsulated into one class

### CODE REVIEW

**Comment by UdjinM6** [comment]
Wow, thanks!

However 0.12.1.x branch is in release mode and it is open for critical fixes only now, so please rebase you branch on top of v0.12.2.x.

**Comment by krychlicki** [comment]
done

**Comment by UdjinM6** [comment]
Errr... I'm not sure what you did exactly but it's a mess now (history wise).
Is that what `git rebase -i v0.12.2.x` produced for you? If so, it's quite surprising tbh... You might want to use it again but this time dropping all unrelated commits i.e. change it to be smth like this:
```
d 6bec60653 Don't add non-current wd's to seen map (#1417)
d b6f18282d Reject payment vote if masternode rank can't be calculated (#1422)
d d0ff35ee5 Fix ProcessVerifyBroadcast (#1423)
d 42aa1bb19 clear trafficgraph on clear button click
d d8de71f07 set default sample height
d 03dfb80d5 reduce available traffic graph ranges, add optimized graph data storage
d e6543b9c7 Don't add non-current wd's to seen map (#1417)
d 3069e0c81 bump to 0.12.1.5 (#1418)
d d8fd73fcd Reject payment vote if masternode rank can't be calculated (#1422)
d 7f4ff495c Fix ProcessVerifyBroadcast (#1423)
pick d27c640de clear trafficgraph on clear button click
pick 87b24a31a set default sample height
pick 10bb55778 reduce available traffic graph ranges, add optimized graph data storage
```

**Comment by krychlicki** [comment]
I've used git rebase v0.12.2.x before. Not sure why this mess in history has happened.
Now I did as you advised.
Thanks for your help!!

**Comment by UdjinM6** [comment]
Np :) history looks good now 👍
Will test first and provide some review feedback after that.

File: src/qt/trafficgraphdata.cpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
+#include<trafficgraphdata.h>
```
Comment (SCDeveloper) [inline src/qt/trafficgraphdata.cpp]: Needs a space.

File: src/Makefile.qttest.include [hunk 2]
```diff
@@ -3,7 +3,8 @@ TESTS += qt/test/test_dash-qt
 
 TEST_QT_MOC_CPP = \
   qt/test/moc_compattests.cpp \
-  qt/test/moc_uritests.cpp
+  qt/test/moc_uritests.cpp \
+  qt/test/moc_trafficgraphdatatests.cpp
```
Comment (UdjinM6) [inline src/Makefile.qttest.include]: sort this alphabetically pls

Comment (UdjinM6) [inline src/Makefile.qttest.include]: sort this alphabetically pls

File: src/qt/test/trafficgraphdatatests.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
+#include "trafficgraphdatatests.h"
+#include "../trafficgraphdata.h"
+#include <algorithm>
+#include <sstream>
+#include <QTime>
+
+void TrafficGraphDataTests::simpleCurrentSampleQueueTests()
+{
+    TrafficGraphData trafficGraphData(TrafficGraphData::Range_5m);
+    for(int i=0; i<TrafficGraphData::DESIRED_DATA_SAMPLES; i++)
```
Comment (UdjinM6) [inline src/qt/test/trafficgraphdatatests.cpp]: code formatting: spaces around "=" and "<" (this applies to many more cases below)

File: src/qt/test/trafficgraphdatatests.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        QVERIFY(trafficGraphData.update(TrafficSample(i,i)));
```
Comment (UdjinM6) [inline src/qt/test/trafficgraphdatatests.cpp]: code formatting: space after "," (this applies to many more cases below)

File: src/qt/test/trafficgraphdatatests.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    QCOMPARE(queue.size(), TrafficGraphData::DESIRED_DATA_SAMPLES);
+    QCOMPARE((int)queue.at(0).in, 0);
+    QCOMPARE((int)queue.at(0).out, 0);
+}
+
+namespace {
+
+void checkQueue(const TrafficGraphData::SampleQueue& queue, int value)
+{
+    for(int i=0;i<queue.size();i++){
```
Comment (UdjinM6) [inline src/qt/test/trafficgraphdatatests.cpp]: code formatting: spaces after ";"

File: src/qt/test/trafficgraphdatatests.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        bool result = trafficGraphData.update(TrafficSample(1,1));
+        std::ostringstream oss;
+        oss<< "result:" << result << " multiplier:" << multiplier << " i:" << i << " range:" << range;
+        if (i==1){
+            if (range == TrafficGraphData::Range_5m)
+                QVERIFY2(result, oss.str().c_str());
+            else
+                QVERIFY2(!result, oss.str().c_str());
+        }
+        else if (i%multiplier == 0)
```
Comment (UdjinM6) [inline src/qt/test/trafficgraphdatatests.cpp]: code formatting: need curly brackets here and in `else` below since `if` already has them

**Comment by krychlicki** [comment]
Thanks!!
Code formatting has been corrected.
Hope it looks ok now.

File: src/qt/trafficgraphwidget.cpp [hunk 7]
```diff
@@ -102,74 +114,52 @@ void TrafficGraphWidget::paintEvent(QPaintEvent *)
         }
     }
 
-    if(!vSamplesIn.empty()) {
-        QPainterPath p;
-        paintPath(p, vSamplesIn);
-        painter.fillPath(p, QColor(0, 255, 0, 128));
+    const TrafficGraphData::SampleQueue& queue = trafficGraphData.getCurrentRangeQueue();
```
Comment (krychlicki) [inline src/qt/trafficgraphwidget.cpp]: I belive spacing is correct here:
const TrafficGraphData::SampleQueue& queue = trafficGraphData.getCurrentRangeQueue();
is inside TrafficGraphWidget::paintEvent method.

Comment (SCDeveloper) [inline src/qt/trafficgraphwidget.cpp]: It's tabbed away from the start of the line.

**Comment by tgflynn** [comment]
The code definitely works well but looking at trafficgraphdata.cpp I find it quite hard to follow due to the complex recursion and lots of special case handling.

Don't you think there are simpler ways to implement this that wouldn't require special case code for the different ranges ?

**Comment by krychlicki** [comment]
The only simpler way is to store all samples for all ranges. 
Then you can just keep it in some collection of queues, push a new one and pop the oldest above sample limit.

In a case when you store only necessary samples it's getting more complicated.
For ranges that I proposed (5m, 10m, 15m, 30m, 1h, 2h, 3h, 6h, 12h, 24h) there are 3 cases:
1) for the smallest range (5m) - all samples need to be stored,
2) for ranges that can be created by taking a sum of 2 samples from smaller range (10m, 30m, 1h, 2h, 6h, 12h, 24h) - 1/2 of samples needs to be stored, the rest is calculated based on sum of samples from smaller range
3) for ranges that can be created by taking a sum of 3 samples from smaller range (15m, 3h) - 2/3 of samples needs to be stored, the rest is calculated based on sum of samples from smaller range

Those 3 cases represent logic for all proposed ranges. 
I used switch statement cointaining every range in order to express clearer what is being done for ech of them.
It could be implemented with switch statement using only 3 cases but I belive I would be harder to understand.
Adtionally sample set for bigger ranges is build based on smaler ranges (eg. for 30m we need to have 15m range, for 15m range we need to have 5m range ) so recursion seams to be natural here.

**Comment by tgflynn** [comment]
The ranges you defined form a sequence where range R_i is k_i*R_(i-1) where the k_i's are small integers (ie. 2 or 3).

For storing the data what if you did something like this (pseudo-code):

`
// R is a vector of range buffers
// K is a vector of subsampling factors
// sampleCounter simply increments for each new sample
addSample(sample) {
  sampleFactor = 1;
  for(i = 0; i < nranges; ++i) {
    if(sampleCounter % sampleFactor == 0) {
      R[i].addSample(sample);
    }
    else {
      break;
    }
    sampleFactor *= K[i];
  }
  ++sampleCounter;
}`

**Comment by krychlicki** [comment]
I can see following problems with this approach:
- defined ranges do not form sequence as you described. eg 15m is k_i*R(i-2)
- i range does not contain only every K[i] sample, it contains a sum of K[i] samples from range R[i-1] or R[i-2]

Additionally, if you don't want to store unnecessary data, you have to add logic that will remove last K[i] add it to R[i] range. Those last K[i] must be samples above R[i-1] or R[i-2] size limit so you need to have some extra space to store them.

**Comment by tgflynn** [comment]
> defined ranges do not form sequence as you described

Sorry, my mistake.

However each range is an integer multiple of the base range.

So we can do this, which is even simpler:

```
// R is a vector of range buffers
// K is a vector of subsampling factors
// sampleCounter simply increments for each new sample

K = {1, 2, 3, 6,12, 24, 36, 72,144, 288};
addSample(sample) {
  for(i = 0; i < nranges; ++i) {
    if(sampleCounter % K[i] == 0) {
      R[i].addSample(sample);
    }
    else {
      break;
    }
  }
  ++sampleCounter;
}
```

As for your second objection I think the easiest way to handle that is to store the cumulative byte counts instead of the instantaneous rates.  You can then compute the rate with 

```
rate = (sampleBuffer[i] - sampleBuffer[i-1])/dt
```

where dt is the timestep for the selected range.

It does require a bit of thought to deal with the edge conditions but I don't think you need extra space since addSample should be storing all the data you need to compute any chosen range.

**Comment by krychlicki** [comment]
ok! that would work.
However, as I described in one of my previous answers this is a simpler approach to store all samples for all ranges.
Here you need to store nranges * DESIRED_DATA_SAMPLES.
I wanted to create a solution that you can store only a minimal amount of samples.
In my approach you need to store:
DESIRED_DATA_SAMPLES + n2ranges * DESIRED_DATA_SAMPLES/2 + n3ranges * DESIRED_DATA_SAMPLES*2/3
where:
n2ranges - number of ranges that are created based on sum of 2 samples from smaller range
n3ranges - based on 3 samples from smaller range

**Comment by tgflynn** [comment]
OK, but what is the rationale for storing fewer samples for some ranges than others ?

Why would you want to store (and display) fewer points for the 10 min range than for the 5 min range, for example ?

**Comment by krychlicki** [comment]
Because the rest we can calculate from other ranges.
for example for 10m range we just need to store half of the needed samples  - second half.
Frist half we can calculate using samples from 5m range.
